### ISSUE TITLE
Redrawing of buttons, scrolled lists etc is slow

### ISSUE THREAD

**Issue by dvtomas** [guillep/pharo-core#5955]
Pharo 8: When I open the Settings Browser, expand a couple of items, and scroll with the mouse wheel, it takes some time before the items that have just appeared in the visible area get painted - subjectively often around 0.5s, which is a lot, looks bad and impairs usability. Is there some kind of redraw throttle somewhere?

**Comment by dvtomas** [comment]
Actually, other actions feel quite choppy as well, even such basic things as pressing a button:

Latest Pharo 9:

I open a big button with 
```smalltalk
b := SpPresenter new newButton.
b icon: (b iconNamed: #pharoBig).
b label: 'Hello, Pharo'.
b openWithSpec.
```

When I press & hold left mouse button, it takes a while until the background changes to indicate it has been pressed. Sometimes, it is almost instant, but sometimes it takes almost one second, which is incredibly long for a smooth UI experience. 

Response for such basic actions should be < 0.1 s. See e. g. https://medium.com/@slhenty/ui-response-times-acec744f3157. Quote: *“0.1 second is the response time limit if you want users to feel like their actions are directly causing something to happen on the screen… Thus, to create the illusion of direct manipulation, a user interface must be faster than 0.1 second.”*

<blockquote><div><img src="https://cdn-static-1.medium.com/_/fp/icons/favicon-rebrand-medium.3Y6xpZ-0FSdWDnPM3hSBIA.ico" height="14"> Medium</div><div><strong><a href="https://medium.com/@slhenty/ui-response-times-acec744f3157">UI Response Times</a></strong></div><div>This article is intended as a reference for defining and characterizing response times in a user interface. The classifications are derived…</div></blockquote>

**Comment by bencoman** [comment]
On Windows 10 with a reasonably recent Pharo 9 
```
Image: Pharo-9.0.0+build.145.sha.435854cd (64 Bit)
VM: Win32 built on Jan  6 2020 16:23:32 CET Compiler: 4.2.1 Compatible Clang 8.0.1 (tags/RELEASE_801/final)
VMMaker versionString VM: 201911290039-dev-7 
```
I experience no perceptible delay in the response.  Your experience seems usual.  I've not heard of such long delays with button on any supported platform (win/lin/osx).  Which OS/version are you on and what are your machine specs?

I presume you experience this in a completely fresh image, but anyway can you check for any green threads running... `World menu > System > Process Browser`
doesn't have anything except these processes...
![image](https://user-images.githubusercontent.com/1713447/77247688-5ff02f00-6c6e-11ea-8a9a-106214ec7c99.png)

Also, what does your OS resource monitor show for CPU/mem?

P.S. For additional visual feedback of pressing the button you also do the following...
```
b action: [ self inform: 'Button pressed' ].
```

**Comment by dvtomas** [comment]
What I try is click and hold LMB on the big button I created above. It changes bg color to blue after a while. The "while" is quite short most of the times, but usually perceptible. In about one in five to ten trials it takes at least 0.5s before it turns blue, but it is hard to guess the exact duration. Maybe the duration is more often longer if I move the mouse before trying, but could be just a coincidence.

System info:

Absolutely fresh Pharo 9 image just obtained via launcher.

Linux Debian 10 (current testing), Intel(R) Core(TM) i5-2400 CPU @ 3.10GHz, plenty of free RAM

Process Browser does indeed show only the mentioned processes.


```
~/opt/pharolauncher/pharo-vm$ ./pharo -version
5.0-201901051900  Sat Jan  5 19:12:50 UTC 2019 gcc 4.8 [Production Spur 64-bit VM]
CoInterpreter VMMaker.oscog-eem.2504 uuid: a00b0fad-c04c-47a6-8a11-5dbff110ac11 Jan  5 2019
StackToRegisterMappingCogit VMMaker.oscog-eem.2504 uuid: a00b0fad-c04c-47a6-8a11-5dbff110ac11 Jan  5 2019
VM: 201901051900 https://github.com/OpenSmalltalk/opensmalltalk-vm.git
Date: Sat Jan 5 20:00:11 2019 CommitHash: 7a3c6b64
Plugins: 201901051900 https://github.com/OpenSmalltalk/opensmalltalk-vm.git
Linux travis-job-f22c8934-2412-48ed-8180-7a42b62c7389 4.4.0-101-generic #124~14.04.1-Ubuntu SMP Fri Nov 10 19:05:36 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
plugin path: ./lib/pharo/5.0-201901051900 [default: /home/td/opt/pharolauncher/pharo-vm/lib/pharo/5.0-201901051900/]
```

**Comment by Ducasse** [comment]
We got some problems because we started to clean some morphic ugly code (such as multiple render loop, event handling,...).
So we are looking for reproducible case.

**Comment by Ducasse** [comment]
I tried your example and I do not get any lag with the latest version of P9.

Build information: Pharo-9.0.0+build.187.sha.cc467ed8847816e888062fec06e6c875dea95fbc (64 Bit)

**Comment by dvtomas** [comment]
Well, I can reproduce this in a couple of seconds on my Pharo installation, just not deterministically. I'm afraid a deterministic reproduction of the behavior might not be possible, the very nature of this problem could be something like "*Depends on how long does it take for the render code to be invoked after the click*" or something like that. That doesn't make it any less a real issue, though.

If there's any way I can help, let me know. Just keep in mind that I'm a Pharo beginner, please.

**Comment by dvtomas** [comment]
What OS did you try on?

**Comment by dvtomas** [comment]
Hmm, OK, I think I've found out the button issue, it is indeed not a Pharo problem. When I monitor the mouse events with `xev`, the ButtonPress event does take some time to arrive by itself. I guess it is the delay necessary for 3rd button emulation (LMB + RMB). If I press & release LMB, the response is instant, both in `xev` and in Pharo, as the 3rd button emulation doesn't apply for press & release. My sincere apologies about that.

The original choppy scrolling problem does persist, however. It is just probably not connected to the button issue in any way, sorry for the false lead.

**Comment by Ducasse** [comment]
Do not worry we prefer that you report a bug that does not exist than the inverse. 

We are working on dropping all the event management of the VM because it is old abd incomplete sometimes we do not get event or no event (and this is why the scrollbar have such a bad behavior) in the near future we will have direct access to os level event via SDL20.

**Comment by dvtomas** [comment]
Should I open a new issue for the original scrolling problem, free of the button clutter? That one is still a problem.

**Comment by dvtomas** [comment]
Just a closing remark, maybe it could help: I wanted to record the behavior of the original bug, where the settings browser, when scrolled, repaints part of the visible area (roughly bottom 20%) with some delay (0.5 s, rough guess). I started recording with gtk-recordmydesktop, and the most amazing thing happened - the redrawing went smoothly, there was no choppines. Stop recording, redrawing starts to be delayed again. Tried three more times, always with the same result.

I'll leave it upon others whether this is useful and worth reopening as another issue.
